# Relatório de Análise 5 

## Tratamento de Dados Faltantes

In [10]:
import pandas as pd 

dados=pd.read_csv('dados/aluguel_residencial.csv', sep=';')
dados.head(10)

Tipo           Bairro  Quartos  Vagas  Suites  Area    Valor  \
0           Quitinete       Copacabana        1      0       0    40   1700.0   
1                Casa  Jardim Botânico        2      0       1   100   7000.0   
2         Apartamento           Centro        1      0       0    15    800.0   
3         Apartamento     Higienópolis        1      0       0    48    800.0   
4         Apartamento     Vista Alegre        3      1       0    70   1200.0   
5         Apartamento         Cachambi        2      0       0    50   1300.0   
6  Casa de Condomínio  Barra da Tijuca        5      4       5   750  22000.0   
7  Casa de Condomínio            Ramos        2      2       0    65   1000.0   
8         Apartamento           Centro        1      0       0    36   1200.0   
9         Apartamento           Grajaú        2      1       0    70   1500.0   

   Condominio  IPTU  
0       500.0  60.0  
1         NaN   NaN  
2       390.0  20.0  
3       230.0   NaN  
4         NaN   NaN  
5       301.0  17.0  
6         NaN   NaN  
7         NaN   NaN  
8         NaN   NaN  
9       642.0  74.0

In [11]:
dados.isnull()

Tipo  Bairro  Quartos  Vagas  Suites   Area  Valor  Condominio   IPTU
0      False   False    False  False   False  False  False       False  False
1      False   False    False  False   False  False  False        True   True
2      False   False    False  False   False  False  False       False  False
3      False   False    False  False   False  False  False       False   True
4      False   False    False  False   False  False  False        True   True
...      ...     ...      ...    ...     ...    ...    ...         ...    ...
22575  False   False    False  False   False  False  False       False  False
22576  False   False    False  False   False  False  False       False  False
22577  False   False    False  False   False  False  False       False  False
22578  False   False    False  False   False  False  False       False  False
22579  False   False    False  False   False  False  False       False   True

[22580 rows x 9 columns]

In [12]:
#ou
dados.notnull()

Tipo  Bairro  Quartos  Vagas  Suites  Area  Valor  Condominio   IPTU
0      True    True     True   True    True  True   True        True   True
1      True    True     True   True    True  True   True       False  False
2      True    True     True   True    True  True   True        True   True
3      True    True     True   True    True  True   True        True  False
4      True    True     True   True    True  True   True       False  False
...     ...     ...      ...    ...     ...   ...    ...         ...    ...
22575  True    True     True   True    True  True   True        True   True
22576  True    True     True   True    True  True   True        True   True
22577  True    True     True   True    True  True   True        True   True
22578  True    True     True   True    True  True   True        True   True
22579  True    True     True   True    True  True   True        True  False

[22580 rows x 9 columns]

In [13]:
dados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22580 entries, 0 to 22579
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Tipo        22580 non-null  object 
 1   Bairro      22580 non-null  object 
 2   Quartos     22580 non-null  int64  
 3   Vagas       22580 non-null  int64  
 4   Suites      22580 non-null  int64  
 5   Area        22580 non-null  int64  
 6   Valor       22571 non-null  float64
 7   Condominio  20765 non-null  float64
 8   IPTU        15795 non-null  float64
dtypes: float64(3), int64(4), object(2)
memory usage: 1.6+ MB


Por essa tabela conseguimos ver que existem 22580 dados no nosso DataFrame, além disso, e que existem valores nulos nas colunas 'Valor','Condominio'e 'IPTU', já que os valores não-nulos contados dessas colunas diferem do valor total de dados 

In [14]:
dados[dados['Valor'].isnull()]

Tipo           Bairro  Quartos  Vagas  Suites  Area  Valor  \
58    Apartamento  Barra da Tijuca        2      1       1    70    NaN   
1492  Apartamento             Leme        2      0       0    75    NaN   
1683         Casa     Campo Grande        3      4       3   363    NaN   
2012  Apartamento         Botafogo        2      0       0    95    NaN   
2034  Apartamento       Copacabana        2      0       0    72    NaN   
4941         Casa     Campo Grande        3      2       1   100    NaN   
8568  Apartamento             Leme        2      0       1    75    NaN   
8947  Apartamento           Glória        3      0       1   135    NaN   
9149  Apartamento            Gávea        3      1       1   105    NaN   

      Condominio   IPTU  
58         970.0   68.0  
1492       878.0    NaN  
1683         NaN    NaN  
2012      1010.0  170.0  
2034       850.0    NaN  
4941         NaN    NaN  
8568       878.0    NaN  
8947       910.0  228.0  
9149       880.0  221.0

Aqui ele nos mostra os dados que possuem os valores nulos!

Vamos apagar todos os dados nulos do Dataframe 

In [15]:
#Uma outra maneira de fazer essa seleção seria:
#dados[dados['Valor'].notnull()]

A=dados.shape[0]
dados.dropna(subset=['Valor'],inplace =True)
B=dados.shape[0]
A-B

9

Agora vamos ver os dados que tem valores nulos novamente:

In [17]:
dados[dados['Valor'].isnull()]

Empty DataFrame
Columns: [Tipo, Bairro, Quartos, Vagas, Suites, Area, Valor, Condominio, IPTU]
Index: []

Está vazio, pois eliminamos todos os dados nulos!

## Tratamento de Dados Faltantes (continuação)

Ainda temos o Condomínio e o IPTU nulos também são um problema 

Porém nem todas as informações de condomínio nulas são erradas,pode ser que a pessoa viva em uma casa

E o IPTU nulo provavelmente foi o dono que esqueceu de reportar 

In [19]:
dados[dados['Condominio'].isnull()].shape[0] # Vamos ver quantos dados possuem o condomínio nulo

1813

Aqui estamos pegando os dados que são do tipo apartamento e que possuem condomínio nulo 

In [22]:
selecao=(dados['Tipo']=='Apartamento') & (dados['Condominio'].isnull()) 

In [23]:
A=dados.shape[0]
dados=dados[~selecao]  # O til irá inverter a variável  booleana 
B=dados.shape[0]
A-B

745

In [24]:
dados[dados['Condominio'].isnull()].shape[0]

1068

Ainda restam 1068 dados de imóveis que não são AP's e que possuem o condomínio nulo, provavelmente são casas

Agora vamos tratar o IPTU

In [28]:
dados=dados.fillna({'Condominio':0, 'IPTU':0})  # Vamos substituir todos os valores nulos por 0  

In [30]:
dados[dados['Condominio'].isnull()].shape[0]

0

In [31]:
dados[dados['IPTU'].isnull()].shape[0]

0

Não existem mais dados nulos!!

In [32]:
dados.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21826 entries, 0 to 22579
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Tipo        21826 non-null  object 
 1   Bairro      21826 non-null  object 
 2   Quartos     21826 non-null  int64  
 3   Vagas       21826 non-null  int64  
 4   Suites      21826 non-null  int64  
 5   Area        21826 non-null  int64  
 6   Valor       21826 non-null  float64
 7   Condominio  21826 non-null  float64
 8   IPTU        21826 non-null  float64
dtypes: float64(3), int64(4), object(2)
memory usage: 1.7+ MB


Agora temos 21826 dados não nulos no nosso DataFrame

In [34]:
dados.to_csv('dados/aluguel_residencial.csv', sep=';',index=False)

In [35]:
range(10)

range(0, 10)